### ISSUE TITLE
certbot-auto incompatible with Python 3.6

### ISSUE THREAD

**Issue by fuchen2624** [certbot/certbot#5456]
## By default , on my operating system ,there is installed python3.4,it produced this output:
Creating virtual environment...
Error: Command '['/opt/eff.org/certbot/venv/bin/python3', '-Im', 'ensurepip', '--upgrade', '--default-pip']' returned non-zero exit status 1

## I was search from google ,It's said there has some bug about python3.4, then I installed a python3.6 on my operating system in  '/usr/local/bin' ,and link python3 to python3.6. then I ran the command  'certbot-auto' again,  it produced this output :
Creating virtual environment...
Installing Python packages...
Had a problem while installing Python packages.

pip prints the following errors: 
Collecting argparse==1.4.0 (from -r /tmp/tmp.fTn8IZl1cS/letsencrypt-auto-requirements.txt (line 11))
  Downloading http://mirrors.aliyun.com/pypi/packages/f2/94/3af39d34be01a24a6e65433d19e107099374224905f1e0cc6bbe1fd22a2f/argparse-1.4.0-py2.py3-none-any.whl
Collecting pycparser==2.14 (from -r /tmp/tmp.fTn8IZl1cS/letsencrypt-auto-requirements.txt (line 17))
  Downloading http://mirrors.aliyun.com/pypi/packages/6d/31/666614af3db0acf377876d48688c5d334b6e493b96d21aa7d332169bee50/pycparser-2.14.tar.gz (223kB)
Collecting asn1crypto==0.22.0 (from -r /tmp/tmp.fTn8IZl1cS/letsencrypt-auto-requirements.txt (line 21))
  Downloading http://mirrors.aliyun.com/pypi/packages/97/ba/7e8117d8efcee589f4d96dd2b2eb1d997f96d27d214cf2b7134ad8acf6ab/asn1crypto-0.22.0-py2.py3-none-any.whl (97kB)
Collecting cffi==1.10.0 (from -r /tmp/tmp.fTn8IZl1cS/letsencrypt-auto-requirements.txt (line 24))
  Downloading http://mirrors.aliyun.com/pypi/packages/99/2f/2304b68ee2eea8395fb98702077e667849b98b9037956fafbef65fd53e0b/cffi-1.10.0-cp36-cp36m-manylinux1_x86_64.whl (406kB)
Collecting ConfigArgParse==0.12.0 (from -r /tmp/tmp.fTn8IZl1cS/letsencrypt-auto-requirements.txt (line 61))
  Downloading http://mirrors.aliyun.com/pypi/packages/17/8d/4a41f11b0971017c7001f118be8003da8f7b96b010c66cd792b76658d1e1/ConfigArgParse-0.12.0.tar.gz (41kB)
Collecting configobj==5.0.6 (from -r /tmp/tmp.fTn8IZl1cS/letsencrypt-auto-requirements.txt (line 63))
  Downloading http://mirrors.aliyun.com/pypi/packages/64/61/079eb60459c44929e684fa7d9e2fdca403f67d64dd9dbac27296be2e0fab/configobj-5.0.6.tar.gz
Collecting cryptography==2.0.2 (from -r /tmp/tmp.fTn8IZl1cS/letsencrypt-auto-requirements.txt (line 65))
  Downloading http://mirrors.aliyun.com/pypi/packages/d9/1c/299fc042685c288ace5d8c0a98474ff8d2e0b95e747eb89be3f93571ef3a/cryptography-2.0.2-cp36-cp36m-manylinux1_x86_64.whl (2.2MB)
Collecting enum34==1.1.2 (from -r /tmp/tmp.fTn8IZl1cS/letsencrypt-auto-requirements.txt (line 96))
  Downloading http://mirrors.aliyun.com/pypi/packages/6f/e9/08fd439384b7e3d613e75a6c8236b8e64d90c47d23413493b38d4229a9a5/enum34-1.1.2.tar.gz (46kB)
    Complete output from command python setup.py egg_info:
    Traceback (most recent call last):
      File "<string>", line 1, in <module>
      File "/opt/eff.org/certbot/venv/lib/python3.6/site-packages/setuptools/__init__.py", line 5, in <module>
        import distutils.core
      File "/usr/local/lib/python3.6/distutils/core.py", line 16, in <module>
        from distutils.dist import Distribution
      File "/usr/local/lib/python3.6/distutils/dist.py", line 9, in <module>
        import re
      File "/usr/local/lib/python3.6/re.py", line 142, in <module>
        class RegexFlag(enum.IntFlag):
    AttributeError: module 'enum' has no attribute 'IntFlag'
    Command "python setup.py egg_info" failed with error code 1 in /tmp/pip-build-0m_n85yu/enum34/

Certbot has problem setting up the virtual environment.

We were not be able to guess the right solution from your pip 
output.

Consult https://certbot.eff.org/docs/install.html#problems-with-python-virtual-environment
for possible solutions.
You may also find some support resources at https://certbot.eff.org/support/ .

**Comment by SwartzCr** [comment]
@erikrose do you think you might be able to help with this?

**Comment by erikrose** [comment]
This looks like our fault: we apparently use `enum34`, a backport of the `enum` module to Python <3.4, and [it's not compatible with Python 3.6](https://stackoverflow.com/a/45716067). My favorite solution is to change `enum34` so it doesn't do anything under newer versions of Python. That's [filed](https://bitbucket.org/stoneleaf/enum34/issues/21/setuppy-should-no-op-for-python-34), but somebody needs to actually do it. It should be pretty easy.

**Comment by erikrose** [comment]
In the meantime, installing under Python 2.7, if you have it, should work.

**Comment by SwartzCr** [comment]
I'm gonna tag @bmw here as well just in case this comes up again or if it leads to us changing this code as we begin migration to more python3 environments

**Comment by erikrose** [comment]
I looked at re.py in the last version of Python 3.5.x, and it's fine. It's only 3.6 that should give us the problem.

**Comment by jcyrss** [comment]
I encountered with same issue, and since our appliaction running with python 3.6, I cannot replace it with the previous version.

Could you tell us when this could be solved?

**Comment by Adoniasv** [comment]
I had the same issue, it looks it was the locale in my case, i just had to run this command:

`export LC_ALL=C`

then try to execute the certbot-auto command again, I hope it helps you.

**Comment by erikrose** [comment]
I've added this to my queue, so I should get to it eventually.

**Comment by erikrose** [comment]
As I graphed out the dependencies of certbot and its default plugins, it seemed like enum34 was used only for the setuptools extras `testing` and `dev`—not suitable for installation by certbot-auto. However, I finally determined that it's pulled in by cryptography, via a setuptools extra where they (strangely?) stowed an environment marker of ":python_version < '3'".

**Comment by erikrose** [comment]
[It's not enough to fix enum34.](https://bitbucket.org/stoneleaf/enum34/issues/21/setuppy-should-no-op-for-python-34#comment-43295808) (We could do some crazy thing where we turn enum34 into just a wrapper package that uses environment markers to install (or choose not to install) a core package that does the work, but…yuck.) It looks like enum34 is pulled into certbot-auto installs only by cryptography[1], so let's fix it there by tightening its env marker. This is why we bother to pipstrap up to a version of pip that understands env markers. :-)

[1] Verified by doing `pip install --no-cache-dir -e /foo/bar/cryptography -e acme -e . -e certbot-apache -e certbot-nginx` where /foo/bar/cryptography is an arbitrary package that calls itself "cryptograph" but lacks an enum34 dep, effectively stubbing out cryptography's effect on the dependency tree. Neither pip's installation chatter nor a subsequent `pip freeze` show the presence of enum34.

**Comment by erikrose** [comment]
Ugh, unfortunately, this does nothing for certbot-auto, which has no affordance for conditional deps anymore. Back to the old drawing board.

**Comment by erikrose** [comment]
Wait, we're [dropping 3.3 support](https://github.com/certbot/certbot/issues/5312)! So, once that's done, we can fix cryptography's env marker, remove enum34 from our certbot-auto pinnings, and we're golden. Phew!

**Comment by jacobbaratta** [comment]
What's the status on a fix? I've had to resort to manually editing the certbot-auto binary to delete the enum34 dependency/hashes in order to get it to work at all with Python 3.6 (SCL) / CentOS 6 (every other attempted fix I tried didn't work), but obviously that means I'm stuck running it with `--no-self-upgrade` now, so I have to manually remember to go update certbot-auto (and then manually repatch it after each update until a proper fix is released), which is rather far from ideal.

**Comment by bmw** [comment]
@erikrose, what needs to be done here and what's the proposed change to `cryptography`? Do they not need `enum34` for Python 2.7? If they do, because we don't have the ability to conditionally install dependencies in `certbot-auto` and I don't see how changing the environment marker helps us.

**Comment by erikrose** [comment]
Indeed, I must have got tunnel vision; it seems like they still need enum34 under 2.7. Solutions I see are…

1. Reintroduce conditional dependencies (my favorite)
2. Update enum34 to actually be enum37 (and keep updating it every time a new Python comes out)
3. Drop Python 2.x support (mentioned for completeness)

**Comment by jacobbaratta** [comment]
@erikrose would installing enum-compat ([PyPI](https://pypi.python.org/pypi/enum-compat), [GitHub](https://github.com/jstasiak/enum-compat)) work as a solution? It's a virtual package that no-ops for Python 3.4+ but otherwise will install enum34. I'm not sure how easy this would be to integrate into certbot-auto given that you explicitly specify all packages and their hashes (making me wonder if this means dependencies must be explicitly specified), but if dependencies don't actually need to be explicitly specified / can be specified inside an installed package, then this would be a solution, right?

The only other alternative I can think of is evaluating python version and switching between two different package lists (one lacking enum34, the other containing it) based on that. It should be doable, but the wholesale duplication of the large dependencies list would be annoying.

**Comment by erikrose** [comment]
enum-compat evaluates a conditional inside setup.py, which happens only at package build time. Thus, any wheels built with this—including ones built implicitly by pip for the local egg cache—won't have the ability to choose at runtime. They could get this working by using an environment marker. However, as you guess, even that won't help us, since we have to pin all (even indirect) dependencies to ensure authenticity.

**Comment by bmoffitt** [comment]
I'm having this problem on a Centos6 server with Python 3.4.5. I tried doing export LC_ALL=C, but the problem persists:
Error: Command '['/opt/eff.org/certbot/venv/bin/python3', '-Im', 'ensurepip', '--upgrade', '--default-pip']' returned non-zero exit status 1

I'm a bit at wit's end, and my certificate hasn't been able to renew and  is expiring tomorrow. Any advice?

**Comment by jacobbaratta** [comment]
@bmoffitt here's the patch I'm using for now:

```
[root@athena certbot-manual]# diff certbot-auto-unmodified certbot-auto
1092,1094d1091
< enum34==1.1.2 \
<     --hash=sha256:2475d7fcddf5951e92ff546972758802de5260bf409319a9f1934e6bbc8b1dc7 \
<     --hash=sha256:35907defb0f992b75ab7788f65fedc1cf20ffa22688e0e6f6f12afc06b3ea501
```

Make the above changes and run certbot-auto with the `--no-self-upgrade` flag (you'll need to periodically manually update it from now on, manually making these changes after every update).

Here's how to interpret the diff format above: https://unix.stackexchange.com/questions/81998/understanding-of-diff-output

And here's the diff in unified format, in case that's easier for you to understand:

```
[root@athena certbot-manual]# diff -u certbot-auto-unmodified certbot-auto
--- certbot-auto-unmodified	2018-04-16 16:30:53.951581610 -0500
+++ certbot-auto	2018-04-16 16:31:12.068866314 -0500
@@ -1089,9 +1089,6 @@
     --hash=sha256:01e6e60654df64cca53733cda39446d67100c819c181d403afb120e0d2a71e1b \
     --hash=sha256:d46f4e5d455cb5563685c52ef212696f0a6cc1ea627603218eabbd8a095291d8 \
     --hash=sha256:3780b2663ee7ebb37cb83263326e3cd7f8b2ea439c448539d4b87de12c8d06ab
-enum34==1.1.2 \
-    --hash=sha256:2475d7fcddf5951e92ff546972758802de5260bf409319a9f1934e6bbc8b1dc7 \
-    --hash=sha256:35907defb0f992b75ab7788f65fedc1cf20ffa22688e0e6f6f12afc06b3ea501
 funcsigs==1.0.2 \
     --hash=sha256:330cc27ccbf7f1e992e69fef78261dc7c6569012cf397db8d3de0234e6c937ca \
     --hash=sha256:a7bb0f2cf3a3fd1ab2732cb49eba4252c2af4240442415b4abce3b87022a8f50
```

**Comment by xtrycatchx** [comment]
i encountered the same prob, and using temporary solution from @Adoniasv solved it. In my case, i just needed to be patient as the process took quite long

**Comment by erikrose** [comment]
@xtrycatchx Assuming you just encountered the problem after a period of not updating, you shouldn't have it again; we fixed it in 0.24.0.
